server/channels/app/authentication.go

```diff
@@ -61,6 +61,64 @@ func (a *App) IsPasswordValid(rctx request.CTX, password string) *model.AppError
 	return nil
 }
 
+// checkUserPassword tries to match the provided password against the user's
+// stored hash, migrating the latter if needed to the latest hashing method.
+func (a *App) checkUserPassword(user *model.User, password string, invalidateCache bool) *model.AppError {
+	err := users.CheckUserPassword(user, password)
+	if err == nil {
+		return nil
+	}
+
+	// Migrate the password if the hashing method is oudated
+	if errors.Is(err, users.OutdatedPasswordHashingError) {
+		return a.migratePassword(user, password)
```

Comment:
I had to dig deep into `migratePassword` to confirm that we actually check that the password was correct before we migrate it. This feels somewhat convoluted: we "check user password" above to find out the hash is wrong, but then we "actually check the user password" in `migratePassword`.

Would it make sense to inline lots of this, and just have the checks be very upfront:
- is password correct relative to existing hash? no, fail early
- if hash up-to-date? no, migrate (no additional check required)